Kernel/VirtIO/VirtIOConsole.h

```diff
@@ -1,52 +1,67 @@
 /*
- * Copyright (c) 2021, the SerenityOS developers.
+ * Copyright (c) 2021, Kyle Pereira <hey@xylepereira.me>
  *
  * SPDX-License-Identifier: BSD-2-Clause
  */
 
 #pragma once
 
-#include <Kernel/Devices/CharacterDevice.h>
 #include <Kernel/VM/RingBuffer.h>
 #include <Kernel/VirtIO/VirtIO.h>
+#include <Kernel/VirtIO/VirtIOConsolePort.h>
 
 namespace Kernel {
+class VirtIOConsole
+    : public VirtIODevice
+    , public RefCounted<VirtIOConsole> {
+    friend VirtIOConsolePort;
 
-#define VIRTIO_CONSOLE_F_SIZE (1 << 0)
-#define VIRTIO_CONSOLE_F_MULTIPORT (1 << 1)
-#define VIRTIO_CONSOLE_F_EMERG_WRITE (1 << 2)
-
-#define RECEIVEQ 0
-#define TRANSMITQ 1
-
-class VirtIOConsole final : public CharacterDevice
-    , public VirtIODevice {
 public:
     VirtIOConsole(PCI::Address);
-    virtual ~VirtIOConsole() override;
+    virtual ~VirtIOConsole() override = default;
 
     virtual const char* purpose() const override { return class_name(); }
 
+    int device_id() const
+    {
+        return m_device_id;
+    }
+
+    void send_open_control_message(Badge<VirtIOConsolePort>, unsigned port_number, bool open) { send_open_control_message(port_number, open); };
+
 private:
-    constexpr static size_t RINGBUFFER_SIZE = 2 * PAGE_SIZE;
-    virtual const char* class_name() const override { return m_class_name.characters(); }
+    enum class ControlEvent : u16 {
+        DeviceReady = 0,
+        DeviceAdd = 1,
+        PortReady = 3,
+        ConsolePort = 4,
+        PortOpen = 6,
+    };
+    struct [[gnu::packed]] ControlMessage {
+        u32 id;
+        u16 event;
+        u16 value;
+    };
 
-    virtual bool can_read(const FileDescription&, size_t) const override;
-    virtual KResultOr<size_t> read(FileDescription&, u64, UserOrKernelBuffer&, size_t) override;
-    virtual bool can_write(const FileDescription&, size_t) const override;
-    virtual KResultOr<size_t> write(FileDescription&, u64, const UserOrKernelBuffer&, size_t) override;
+    const static int CONTROL_RECEIVEQ = 2;
+    const static int CONTROL_TRANSMITQ = 3;
+    const static size_t CONTROL_BUFFER_SIZE = 256;
 
-    virtual mode_t required_mode() const override { return 0666; }
+    const char* class_name() const { return m_class_name.characters(); }
```

Comment:
this functionality should just be moved into `purpose` now that this is not a virtual override anymore.